### ISSUE TITLE
Are Turtle.Prelude.stream and Turtle.Prelude.system exception safe?

### ISSUE THREAD

**Issue by massysett** [Gabriel439/Haskell-Turtle-Library#34]
I noticed they acquire resources (the process and its associated file handles) but they don't use something like Managed.  Is that something to be concerned about?  (This is more a "question" than an "issue".)

**Comment by Gabriel439** [comment]
When I wrote them up I didn't even realize there was a way to dispose of processes in the event of an exception (using `terminateProcess`).  It also didn't occur to me that the `Handle`s needed to be disposed of because I didn't make the mental connection between process handles and file handles.

So the answer is that they are probably going to at least leak `Handle`s even when there are no exceptions.  It's not clear from the documentation if it's possible to leak a process, but it can't hurt to clean that up anyway, too.  Both of those are something I need to fix.

**Comment by massysett** [comment]
Check out this commented-out code from System.Process.  Why they are not confident enough to uncomment out this code, I do not know.

{-
-- TODO: decide if we want to expose this to users
-- | A 'C.bracketOnError'-style resource handler for 'createProcess'.
--
-- In normal operation it adds nothing, you are still responsible for waiting
-- for (or forcing) process termination and closing any 'Handle's. It only does
-- automatic cleanup if there is an exception. If there is an exception in the
-- body then it ensures that the process gets terminated and any 'CreatePipe'
-- 'Handle's are closed. In particular this means that if the Haskell thread
-- is killed (e.g. 'killThread'), that the external process is also terminated.
--
-- e.g.
--
-- > withCreateProcess (proc cmd args) { ... }  $ \_ _ _ ph -> do
-- >   ...
--
withCreateProcess
  :: CreateProcess
  -> (Maybe Handle -> Maybe Handle -> Maybe Handle -> ProcessHandle -> IO a)
  -> IO a
withCreateProcess c action =
    C.bracketOnError (createProcess c) cleanupProcess
                     (\(m_in, m_out, m_err, ph) -> action m_in m_out m_err ph)
-}

**Comment by Gabriel439** [comment]
So I'm going to leave this unfixed for now since I'm not totally sure if it's a problem yet.  However, I'll leave the issue open in case somebody has an issue of leaking too many `Handle`s.

**Comment by 3noch** [comment]
Have you confirmed that there is a handle leak in general?

**Comment by Gabriel439** [comment]
Yes, I was able to confirm that this leaks handles if run repeatedly, even if there are no exceptions.  The easiest way to reproduce is to run this script:

```haskell
#!/usr/bin/env runhaskell

{-# LANGUAGE OverloadedStrings #-}

import Turtle

main = sh (do
    select (repeat ())
    liftIO (shell "true" empty) )
```

... and then monitor the number of open handles using `lsof`.

I'll fix these commands to ensure that the `Handle`s are properly closed afterwards (both with and without exceptions).

**Comment by Gabriel439** [comment]
Hmmm.  For some really strange reason once I start `bracket`ing the computations the commands hang indefinitely when fed input.  I believe `process` is probably doing something really weird and unsafe.

I may have to leave this issue open until I figure out what is wrong with the `process` library

**Comment by PierreR** [comment]
I am not sure this is related but while doing the example from the `slides.md` I have been very surprised by this behavior (within ghci):

```
> stdout $ input "dontexist"
*** Exception: buld.sh: openFile: does not exist (No such file or directory)
-- No error !!
> stdout $ input "dontexist" & inshell "nl"
```

**Comment by Gabriel439** [comment]
I fixed them so that they are now exception-safe and they also properly clean-up after themselves so that handles don't leak.  This was fixed by aef155925b8f4c70c69366e78c4e04c0decb4fd8
